### PR TITLE
Utility extensions for master-koin

### PR DESCRIPTION
**Pull request by hrodrick**
#### Description
* Added some utilty extensions and classes that can be used in any project to reduce the amount of code needed to do some tasks and to not re-engineer them again. I've used them in a lot of projects so hope they will be useful for future devs! 
* This is the same PR as https://github.com/rootstrap/android-base/pull/59 but aims to be merged into the master-koin template

#### Preview
* There is a `CustomTypeFaceSpan` that allows to set a custom font to parts of a text. 
* Added a `Fragment.collectOnLifeCycle(...)` extension to reduce the boiler plate code and annoying indentation when collecting flows from a fragment. (It can be easily adapted to activities)
* Added a `ProgressBar.progressTo(...)` extension to animate progress updates in one line with ease and good support for older android versions.
* Added a `TextView.setClickableKeyword(...)` extension to add clickable functions to words or prhases inside a Textview, allowing to also change their color, font (for example, to bold the keyword), and underline. It recycles the LinkMovementMethod if the TextView already have it.
* Added a `TextView.setColoredKeyword(...)` to change the color of a word or prhase to a Textview.
These two last functions uses the `getTextAsSpannable()` private extension to avoid creating unuseful spannables.

### CODE REVIEW

File: app/src/main/java/com/rootstrap/android/ui/custom/CustomTypefaceSpan.kt [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        private fun applyCustomTypeFace(paint: Paint, tf: Typeface) {
+            val oldStyle: Int
+            val old = paint.typeface
+            oldStyle = old?.style ?: 0
+            val fake = oldStyle and tf.style.inv()
+            if (fake and Typeface.BOLD != 0) {
+                paint.isFakeBoldText = true
+            }
+            if (fake and Typeface.ITALIC != 0) {
+                paint.textSkewX = -0.25f
```
Comment (agustinkoll-rootstrap) [inline app/src/main/java/com/rootstrap/android/ui/custom/CustomTypefaceSpan.kt:34]: What's the meaning of 0.25 ? can be a constant ?

File: app/src/main/java/com/rootstrap/android/util/extensions/Fragment.kt [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ * you will need to cast the flow values when needed.
+ *
+ * Example of use : <br>
+ * - Single flow collection :
+ *      > `collectOnLifeCycle(Pair(viewModel.eventsFlow) { it.collect { event -> onEvents(event) } })`
+ * - Multi flow collection :
+ *      > Just appending a trailing "," and adding a new Pair allows you to collect an additional
+ *          flow per pair
+ */
+fun <T> Fragment.collectOnLifeCycle(
```
Comment (agustinkoll-rootstrap) [inline app/src/main/java/com/rootstrap/android/util/extensions/Fragment.kt:30]: This is cool

File: app/src/main/java/com/rootstrap/android/ui/custom/CustomTypefaceSpan.kt [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import android.text.TextPaint
+import android.text.style.TypefaceSpan
+
+/**
+ * Allows to apply a custom font in an spannableString
+ *
+ * Taken from https://stackoverflow.com/questions/10675070/multiple-typeface-in-single-textview
+ * with slightly modifications
+ */
+class CustomTypefaceSpan(private val newType: Typeface) : TypefaceSpan("") {
```
Comment (amaury901130) [inline app/src/main/java/com/rootstrap/android/ui/custom/CustomTypefaceSpan.kt:14]: I'd rename to: `TypefaceSpan`

File: app/src/main/java/com/rootstrap/android/util/extensions/ProgressBar.kt [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+fun ProgressBar.progressTo(
+    newlyProgress: Int,
+    timeInMillis: Long = DEFAULT_ANIMATION_TIME,
+    onEndListener: ((Animator) -> Unit)? = null
+) {
+    val animation: ObjectAnimator = ObjectAnimator
+        .ofInt(this, "progress", progress, newlyProgress)
+    animation.duration = timeInMillis
+    animation.interpolator = DecelerateInterpolator()
+    animation.setAutoCancel(true)
```
Comment (amaury901130) [inline app/src/main/java/com/rootstrap/android/util/extensions/ProgressBar.kt:23]: use scope function here.

File: app/src/main/java/com/rootstrap/android/util/extensions/Textview.kt [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+import android.text.style.CharacterStyle
+import android.text.style.ClickableSpan
+import android.view.View
+import android.widget.TextView
+import androidx.annotation.ColorRes
+import androidx.core.content.ContextCompat
+import com.rootstrap.android.R
+import com.rootstrap.android.ui.custom.CustomTypefaceSpan
+
+fun TextView.setClickableKeyword(
```
Comment (amaury901130) [inline app/src/main/java/com/rootstrap/android/util/extensions/Textview.kt:17]: add a description here.

File: app/src/main/java/com/rootstrap/android/util/extensions/Textview.kt [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    span.setSpan(clickableSpan, start, start + keyword.length, Spanned.SPAN_INCLUSIVE_INCLUSIVE)
+
+    if (movementMethod !is LinkMovementMethod) {
+        movementMethod = LinkMovementMethod.getInstance()
+    }
+
+    text = span
+}
+
+fun TextView.setColoredKeyword(
```
Comment (amaury901130) [inline app/src/main/java/com/rootstrap/android/util/extensions/Textview.kt:48]: same

**Review by amaury901130 — APPROVED** [review]
Looking good, made sure to add a mini description of these extensions in the readme file.

File: app/src/main/java/com/rootstrap/android/ui/custom/CustomTypefaceSpan.kt (hunk 1 above)
Comment (hrodrick) [reply app/src/main/java/com/rootstrap/android/ui/custom/CustomTypefaceSpan.kt:34]: Yeah it is the skew factor needed to simulate an italic typeface. You can think of this as the angle the letters are moved to _look like this_

File: app/src/main/java/com/rootstrap/android/util/extensions/Textview.kt (hunk 5 above)
Comment (hrodrick) [reply app/src/main/java/com/rootstrap/android/util/extensions/Textview.kt:17]: Added a KDoc with an example usage 👍🏼

File: app/src/main/java/com/rootstrap/android/util/extensions/Textview.kt (hunk 6 above)
Comment (hrodrick) [reply app/src/main/java/com/rootstrap/android/util/extensions/Textview.kt:48]: Added a KDoc here too. Thanks! 👍🏼

**Review by agustinkoll-rootstrap — APPROVED** [review]

**Pull request merged**